### ISSUE TITLE
Support for detecting missing/extraneous object properties in TypeScript

### ISSUE THREAD

**Issue by dstaley** [gajus/eslint-plugin-jsdoc#491]
I'm working through updating the JSDoc comments in one of my libraries, and the doc comments are significantly diverged from the code. I was hoping this plugin would help me catch those instances, and while it's incredibly useful, it's missing one use case that'd be a huge help in my codebase.

Given the following TypeScript, I was hoping eslint would report two errors:

```typescript
/**
 * Testing
 *
 * @param options
 * @param options.one One
 * @param options.two Two
 * @param options.four Four
 */
function testingEslint(options: {
  one: string;
  two: string;
  three: string;
}): string {
  return one + two + three;
}
```

Expected errors:

1. Missing `@param` for `options.three`
2. `@param` "options.four" does not match an existing function parameter

I tested by enabling all the available rules, but please let me know if this is doable through some extra configuration.

**Comment by brettz9** [comment]
Yeah, this is a dupe of #11 . I think it is probably the most useful rule we're still missing (note that that issue also tracks for the similar cases of array destructuring and rest parameters that we are not checking either). Possible to do but just needs some time.

(If anyone is motivated to work on it, I have made a little start at it in my "destructuring" branch, in case that might help point some of the way toward finishing it; I am not sure about getting back to it myself as I have a lot on my plate now.)

Closing, though feel free to follow on #11.

**Comment by dstaley** [comment]
@brettz9 I've successfully managed to adjust `checkParamNames` in such as way that the following is now an error:

```js
/**
 * @param foo
 * @param foo.bar
 */
function quux ({foo, bar}) { }
// Missing @param "foo.foo"
```

However, the above mentioned TypeScript is subtly different. It's not actual destructuring, so the Node type isn't `ObjectPattern`. I believe the AST type would be `TSTypeAnnotation`, and it doesn't look like there's anything in the codebase that handles that type already.

Would you mind pointing me in the right direction on where would be the most appropriate place to implement a check that would fail on the code in the original issue? It's totally possible that it would require a completely new rule, but I'm hoping there's something existing that I can hook into.

If it ends up being pretty complicated to add, I'll send a PR with my existing updates, otherwise I'll hold off until we can get this particular use case working.

**Comment by brettz9** [comment]
I think `check-param-names` is the right rule for this. I don't know if other rules ought to have special checks for `TSTypeAnnotation`, such as `require-param`, but I would expect the handling could be added within the same utilities handling destructuring, i.e., `getFunctionParameterNames` in `jsdocUtils.js` which sets the type, and then any rules which consume `<ObjectPattern>` and the like ought to be checked to see whether returning `<TSTypeAnnotation>` would make sense (I'd need a closer look to be sure, but I suspect that would be a consistent approach).

If needed, new utilities can be added too, or the old methods modified incrementally, so as long as we don't have test/coverage breaking, I'd expect we could move forward.

(If you weren't aware, https://astexplorer.net/ is awesome for debugging these kinds of things, and the parser (i.e., the `</>` pulldown) that you want for TypeScript support is `@typescript-eslint/parser` (as with eslint, we support any eslint-expressed parsers, whether the default espree, or, for `settings.jsdoc.mode === "typescript"` then `@typescript-eslint/parser`). By pasting in your code, you can see there that `TSTypeAnnotation` is indeed the right type and you can also use it to find the expected child properties, etc.

**Comment by brettz9** [comment]
@dstaley : You may want to take a look at #584 both because I think it should fully resolve the cases which had not been fully resolved and also adds reporting for a mismatch between TS annotation names and `@param` root names (as I think makes sense, though I am not a regular TS user).
